consensus/src/payload_manager.rs

```diff
@@ -581,19 +591,41 @@ async fn request_txns_from_quorum_store(
     Ok(ret)
 }
 
-async fn process_payload_helper<T: TDataInfo>(
+async fn process_optqs_payload<T: TDataInfo>(
     data_ptr: &BatchPointer<T>,
     batch_reader: Arc<dyn BatchReader>,
     block: &Block,
     ordered_authors: &[PeerId],
+    additional_peers_to_request: Option<&BitVec>,
 ) -> ExecutorResult<Vec<SignedTransaction>> {
-    let (iteration, fut) = {
+    let (iteration, fut, existing_responders) = {
         let data_fut_guard = data_ptr.data_fut.lock();
         let data_fut = data_fut_guard.as_ref().expect("must be initialized");
-        (data_fut.iteration, data_fut.fut.clone())
+        (
+            data_fut.iteration,
+            data_fut.fut.clone(),
+            data_fut.responders.clone(),
+        )
     };
 
+    let mut signers = Vec::new();
+    if let Some(peers) = additional_peers_to_request {
+        for i in peers.iter_ones() {
+            if let Some(author) = ordered_authors.get(i) {
+                signers.push(*author);
+            }
+        }
+    }
+
+    // Append the additional peers to existing responders list
+    // NB: this might append the same signers multiple times, but this
+    // should be very rare and has no negative effects.
+    for responders in existing_responders {
+        responders.lock().append(&mut signers.clone());
```

Comment:
it feels awkward, every retry will append duplicate signers into the responders?

Reply:
Yeah, the alternate is very verbose. I need to check if the vector is a subset and then add them efficiently. There is no harm re-adding though. It's appended at the end, so it's like effectively cycling over the original vector?

Reply:
I thought the vector grows infinitely with the iterations, but seems it can at most duplicate once?

Reply:
It can grow if the fetch keeps failing forever.

Reply:
I thought every iteration it starts with the qc signers + batch signers? or do i missing something?
```
                    let mut signers = signers.clone();
                    signers.append(&mut summary.signers(ordered_authors));
```

Reply:
Actually I take it back, yes it will only append once. Because the next time its requested I recreate the vector.

Reply:
the code reads very confusing to me, I see that the `batches_and_responders` is re-created but also I don't see the data_ptr.responders do the same, so the data_ptr.responders seems keep growing?

Reply:
🤦🏽‍♂️ yeah you are right. I am not replacing data_ptr.responders. Agree, it is confusing.

Reply:
May be I should I just re-issue the request instead of trying to send the updated peers via the Arc<Mutex<>> and accept that there might be duplicate fetches on QS side. This should be rare anyhow? That way this might be much more cleaner.

Reply:
Orthogonally, may be we should make QS fetcher smarter and consolidate fetches to same batches instead of duplicating concurrent fetches.

Reply:
Okay, I am updating the responders too properly now. Still not straight-forward code. The real fix is to change batch_requester and BatchReaderImpl to efficiently handle duplicate get requests. That way payload manager code can be very minimal, just fetch without caring about tracking previous requests. I can do that in the next PR. wdyt?

Reply:
fine in next pr, probably we should get both into next branch cut